### PR TITLE
implemented sizeof operator and added a feature test

### PR DESCRIPTION
**Pull request by pougetat**

### CODE REVIEW

**Comment by ShivamSarodia** [comment]
Hi pougetat! Thank you for your PR. I have a few suggested changes before it's ready to merge :)

* Firstly, feel free to add yourself as a contributor on the README if you'd like!
* I recommend taking a look at 6.5.3.4 in the C11 spec (see link at bottom of the README) to make sure your PR implements all the nuances of the sizeof operator. Below are a couple concrete points of difference I noticed between the spec and your implementation:
    *  Per 6.5.3.4.1, sizeof can only be applied on certain types/objects with type.
    *  Per 6.5.3.4.5, sizeof the result of sizeof has unsigned type. On my 64-bit machine, GCC appears to implement the result of sizeof as a unsigned long, so it may be best to emulate that.
    *  Per 6.5.3.4.1, sizeof can be also be applied on a type, not just an object. Implementing this requires a small change in the parser.
    * Per 6.5.3.4.1, sizeof does not evaluate the operand sizeof in most cases. This is relevant in a situation like `sizeof(f())` where evaluating the operand would generate ASM code where there should be none. This can be implemented by passing a dummy copy of the IL code object to the make_il function of the sizeof expression. (Make sure if you declare a struct in the sizeof, though, that the struct remains in scope outside the sizeof expression.)
* Could you expand the breadth of your feature test by testing other common and edge cases? For example, test the sizeof an constant expression, sizeof an array, sizeof a whole struct, etc. Rather than print out the test result, it is best to compare a computed result against a known value and return 0 if the result is as expected, and a non-zero number if the result is not what was expected. See feature_tests/addition.c for an example of this.

Thanks again for your contribution - let me know if anything I said was unclear of if you'd like further guidance in implementing any of these suggestions.

**Comment by pougetat** [comment]
Hi Shivam!
Thanks for the review. I will be going over your comments as soon as possible. :)

**Comment by pougetat** [comment]
Hi Shivam!

I'm currently working on the changes you requested and more specifically the sizeof(type) case. Initially I thought that it would indeed require a small modification to the parser but now I am starting to second guess it.
Here is what I was trying to do :
- I wanted to implement an extra method in the parser that would handle just the sizeof case. 
   * If the contents of sizeof are a known type (ex : int) then I return a Paren node containing a Number node with the value of the number being the size of that type. 
   * Otherwise I call the parse_cast method and return whatever it returns (a Paren node)
First of all, the reason I return a Paren node in the first case is that in my sizeof class the call to make_il can operate on the assumption that self.expr is a Paren node therefore making the code more uniform since this is the node returned by parse_cast.
- Now coming back to the case where sizeof contains a known type. I was initially thinking of a bunch of if statements checking if the token is int, char, etc but I realize this is not a good idea since types can be more complex (long int) and therefore many if statements would be needed. To ease the parsing of this potential scenario of a complex type (long int), I was at least thinking of reusing the parse_decl_specifiers method defined in declaration.py but this leads to a circular dependency.

My question is the following :
- Is my current solution (or at least the shape of it) needlessly complex and I am missing something far easier to implement or does this specification (sizeof applied to known types) require a larger modification to the parser than just a 2 line fix. :)
Thanks!

**Comment by ShivamSarodia** [comment]
Thanks for your question! When making parsing decisions, I usually consult the C11 spec (see link at bottom of README) closely because I've found the syntax it provides tends to be generally good guidance on how to implement a parser.

Looking at 6.5.3, the two relevant rules are:
```
unary-expression:
  sizeof unary-expression
  sizeof(type-name)
```
Fortunately, we already have a `parse_unary_expression` function to parse the `unary-expression` in the first rule. The second rule is a little trickier. From 6.7.7, we see that a `type-name` is a `specifier-qualifier-list` followed by something called an abstract declarator. When you look further into it, it turns out that an abstract declarator is like a regular declarator, except without an identifier.

Declaration parsing has been implemented in the parser/declaration.py file. A `specifier-qualifier-list` is a lot like a `declaration-specifiers` (see 6.7), and we can parse those using the `parse_decl_specifiers` function like you suggested (but make sure to check, probably in the `make_il` function, that the list returned is only type specifiers and type qualifiers and nothing else because the `parse_decl_specifiers` function will also accept a `storage-class-specifier`). I wasn't sure what you meant by that function introducing a circular dependency, so let me know if you're still wondering about that. To parse an abstract declarator, you can use the `parse_declarator` function (see the `parse_func_definition` function in parser/declaration.py for an example, since the function call is a touch different than the other `parse_*` functions). The `parse_declarator` function will return a declaration tree, and you can check that tree in your `make_il` function to confirm it doesn't have an identifier in it (that is, you want it to be like `int` or `int[10]`, not `int a` or `int a[10]`).

To summarize: when the parser sees a sizeof, it can first try parsing a unary expression. If that succeeds, you're done! If that fails, catch the ParserError and save it with log_error (you can find this technique used in the parser in a few places). Then, try to match a parenthesis token, and then try to parse a type-name. This consists of using `parse_decl_specifiers` to parse all the specifiers, then using `parse_declarator` to parse the abstract declarator itself. Make sure you match the final parenthesis token too. In the `make_il` function of your sizeof node, check that the declaration specifiers are all legal and that the declarator you parsed doesn't have an identifier in it.

I know that was a ton of jargon I just dropped on you, so please let me know if any of this doesn't make sense! It sounds scary but I believe the implementation of the parsing will only be about 10-15 lines at most. Hope this wasn't too confusing :)

**Comment by ShivamSarodia** [comment]
Hi pougetat, if you haven't implemented the type-name parsing yet feel free to hold off on that for about a day. I'm implementing casts right now, and I think the code I'll write for that will include a function that parses type names, which you can just use directly for your PR.

**Comment by ShivamSarodia** [comment]
Okay, I finished implementing casts which you can see here: https://github.com/ShivamSarodia/ShivyC/blob/33658832970ee94dbf13d966ec81fd33bc4fc5a6/shivyc/parser/expression.py#L119. I didn't actually create a separate function to parse type names, because it is only 2 lines to do so.

The cast operation is a parenthesis, followed by a specifier-qualifier list, followed by an abstract declarator, followed by an end parenthesis. You can see these four lines of code in the link I sent above - hope that helps! 

Let me know if you still have questions or want to run any thoughts by me.

**Comment by ShivamSarodia** [comment]
Oh, also please feel free to add yourself to the Contributors list in the README :)

**Comment by pougetat** [comment]
Hey Shivam,

Thanks for the modifications and the comments. I should be pushing my code very soon. :)
I did notice a bug however when I tried to use ctype.unsig_longint for the il_value instead of ctype.integer.

The error i got was : 
   sizeof.s: Assembler messages:
   sizeof.s:12: Error: operand type mismatch for `cmp'
   shivyc: error: assembler returned non-zero status

I am going to look into this but I just thought I would give you a heads up since this seems to be more backend-related and you may know what this is due to. ;) 
Thanks!

**Comment by pougetat** [comment]
Oh and I will add the tests tomorrow :)

**Comment by ShivamSarodia** [comment]
Thanks for letting me know about the error! I'm having some trouble reproducing it on my computer, so could you please provide a c file produces the error when you get a chance?

Looks like what is happening is the compiler is emitting a `cmp` assembly command with operands that are somehow incorrect for the assembler.

**Comment by pougetat** [comment]
Hey Shivam,
thanks for the response. If you clone my fork and test it on the following C file you might get the error (I have it). =)

int main() {
  sizeof(int) == 19;
  return 0;
}

**Comment by pougetat** [comment]
Hey Shivam,
I made the changes and added a test file. :)

File: tests/feature_tests/sizeof.c [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#include <stdio.h>
+
+int main() {
```
Comment (ShivamSarodia) [inline tests/feature_tests/sizeof.c]: Thank you for adding this very comprehensive test! Just one more thing I thought of - could you add a couple lines to make sure `sizeof` works correctly on a typedefed type, and also that a struct type declared inside a sizeof is visible outside the sizeof function? I'm thinking something like:
```
typedef int T;
if(sizeof(T) != 4) return 1;
  
T a;
if(sizeof(a) != 4) return 2;
  
sizeof(struct S {int a;});
struct S s;
```

**Comment by ShivamSarodia** [comment]
Oh, one other thing - recall that sizeof "shall not be applied to an expression that has function type or an incomplete type, to the parenthesized name of such a type" so we should raise an error in that case.
